File: graph/cpp/kruskals.cpp
#include<bits/stdc++.h>
using namespace std;
struct node{
    int u,v,w;
    node(int first,int second,int weight){
        u=first;
        v=second;
        w=weight;
    }
};
bool comp(node a,node b){
    return a.w < b.w;
}
int findPar(int u,vector<int> &parent){
    if(u == parent[u])
        return u;
    return parent[u] = findPar(parent[u],parent);
}
void Union(int u,int v,vector<int> &parent,vector<int> &size){
    u = findPar(u,parent);
    v = findPar(v,parent);
    if(size[u] < size[v]){
        parent[u] = v;
        size[v] += size[u];
    }else{
        parent[v]=u;
        size[u]+=size[v];
    }
}
int main(){
    int N,m;
    cin>>N>>m;
    vector<node> edges;
    for(int i=0;i<m;i++){
        int u,v,w;
        cin>>u>>v>>w;
        edges.push_back(node(u,v,w));
    }
    /*
        sample testcase for trying
        N=4,m=5
        0 1 10
        0 2 6
        0 3 5
        1 3 15
        2 3 4
    */
    sort(edges.begin(),edges.end(),comp);
    vector<int> parent(N);
    for(int i=0;i<N;i++)
        parent[i] = i;
    vector<int> size(N,0);
    int cost=0;
    vector<pair<int,int>> mst;
    for(auto it:edges){
        if(findPar(it.v,parent) != findPar(it.u,parent)){
            cost+=it.w;
            mst.push_back({it.u,it.v});
            Union(it.u,it.v,parent,size);
        }
    }
    cout<<cost<<endl;
    for(auto it:mst){
        cout<<it.first<<" "<<it.second<<endl;
    }
    return 0;
}

File: graph/python/kruskals.py
class DisjointSet:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n  # Rank optimization
    
    def find(self, v):
        if self.parent[v] != v:
            self.parent[v] = self.find(self.parent[v])  # Path compression
        return self.parent[v]

    def union(self, u, v):
        root1 = self.find(u)
        root2 = self.find(v)

        if root1 != root2:
            if self.rank[root1] > self.rank[root2]:
                self.parent[root2] = root1
            elif self.rank[root1] < self.rank[root2]:
                self.parent[root1] = root2
            else:
                self.parent[root2] = root1
                self.rank[root1] += 1

def kruskal(n, edges):
    edges.sort(key=lambda x: x[2])  # Sort edges by weight
    ds = DisjointSet(n)
    mst = []
    total_weight = 0

    for u, v, weight in edges:
        if ds.find(u) != ds.find(v):  # If adding edge does not form a cycle
            ds.union(u, v)
            mst.append((u, v, weight))
            total_weight += weight

            if len(mst) == n - 1:  # MST complete
                break

    return mst, total_weight

# Example Usage
edges = [
    (0, 1, 10), (0, 2, 6), (0, 3, 5),
    (1, 3, 15), (2, 3, 4)
]
n = 4  # Number of nodes

mst, weight = kruskal(n, edges)
print("MST Edges:", mst)
print("Total MST Weight:", weight)
